spopt/locate/base.py

```diff
@@ -0,0 +1,406 @@
+from abc import abstractmethod
+
+from spopt.BaseClass import BaseSpOptSolver
+from typing import TypeVar
+
+import numpy as np
+import pulp
+
+
+class LocateSolver(BaseSpOptSolver):
+    """
+    Base Class for locate package
+    """
+
+    def __init__(self, name: str, problem: pulp.LpProblem):
+        self.name = name
+        self.problem = problem
+        self.aij = np.array([[]])
+
+    @abstractmethod
+    def solve(self, solver: pulp.LpSolver):
+        """
+        Solve the optimization model
+
+        Parameters
+        ----------
+        solver: pulp.LpSolver
+                solver supported by pulp package
+
+        Returns
+        -------
+        None
+        """
+        pass
+
+
+class BaseOutputMixin:
+    """
+    Base Mixin used by all models
+    """
+
+    def client_facility_array(self) -> None:
+        """
+        Create an array 2d $m$ x $n$, where m is number of clients and n is number of facilities.
+        """
+        if hasattr(self, "fac2cli"):
+            self.cli2fac = [[] for i in range(self.aij.shape[0])]
+
+            for i in range(len(self.fac2cli)):
+                for fac_site in self.fac2cli[i]:
+                    self.cli2fac[fac_site].append(i)
+
+    def uncovered_clients_dict(self) -> None:
+        """
+        Calculate how many clients points are not covered
+        """
+        set_cov = set()
```

Comment:
Let's try get as much code coverage as possible:
* all classes, methods, functions should have at least one test
* testing for known failures is important, too (see comment in `base.FacilityModelBuilder.add_set_covering_constraint()`)